src/System.Linq.Expressions/src/System/Runtime/CompilerServices/CallSite.cs

```diff
@@ -667,7 +667,7 @@ private T CreateCustomNoMatchDelegate(MethodInfo invoke)
                 Expression.Block(
                     Expression.Call(
                         typeof(CallSiteOps).GetMethod(nameof(CallSiteOps.SetNotMatched)),
-                        @params.First()
+                        @params[0]
```

Comment:
Is there a path where that can happen? The only one I traced had gone through line 336:
```C#
ParameterInfo[] pis = invoke.GetParametersCached();
ContractUtils.Requires(pis.Length > 0 && pis[0].ParameterType == typeof(CallSite), nameof(T));
```
So it would have AOoRE already.

If there's another path it that does allow zero parameters then replacing `ParameterExpression[] @params` with `IList<ParameterExpression> @params` in 665 would force the correct exception.